pkg/tools/prefix_pool/prefixpool.go

```diff
@@ -0,0 +1,642 @@
+			prefix, leftPrefixes, err := ExtractPrefix(newPrefixes, request.PrefixLen)
+			if err != nil {
+				// It seems there is no more prefixes available, but since we have all Required already we could go.
+				break
+			}
+			result = append(result, prefix)
+			newPrefixes = leftPrefixes
+		}
+	}
+	if len(result) == 0 {
+		return nil, prefixes, errors.Errorf("Failed to extract prefixes, there is no available %v", prefixes)
+	}
+	return result, newPrefixes, nil
+}
+
+func ExtractPrefix(prefixes []string, prefixLen uint32) (string, []string, error) {
+	// Check if we already have required CIDR
+	max_prefix := 0
+	max_prefix_idx := -1
+
+	// Check if we already have required prefix,
+	for idx, prefix := range prefixes {
+		_, netip, err := net.ParseCIDR(prefix)
+		if err != nil {
+			continue
+		}
+		parentLen, _ := netip.Mask.Size()
+		// Check if some of requests are fit into this prefix.
+		if prefixLen == uint32(parentLen) {
+			// We found required one.
+			resultPrefix := prefixes[idx]
+			resultPrefixes := append(prefixes[:idx], prefixes[idx+1:]...)
+			// Lets remove from list and return
+			return resultPrefix, resultPrefixes, nil
+		} else {
+			// Update minimal root for split
+			if uint32(parentLen) < prefixLen && (parentLen > max_prefix || max_prefix == 0) {
+				max_prefix = parentLen
+				max_prefix_idx = idx
+			}
+		}
+	}
+	// Not found, lets split minimal found prefix
+	if max_prefix_idx == -1 {
+		// There is no room to split
+		return "", prefixes, errors.Errorf("Failed to find room to have prefix len %d at %v", prefixLen, prefixes)
+	}
+
+	resultPrefixRoot := prefixes[max_prefix_idx]
+	right_parts := []string{}
+
+	_, rootCIDRNet, _ := net.ParseCIDR(resultPrefixRoot)
+	for {
+		rootLen, _ := rootCIDRNet.Mask.Size()
+		if uint32(rootLen) == prefixLen {
+			// we found required prefix
+			break
+		}
+		sub1, err := subnet(rootCIDRNet, 0)
+		if err != nil {
+			return "", prefixes, err
+		}
+
+		sub2, err := subnet(rootCIDRNet, 1)
+		if err != nil {
+			return "", prefixes, err
+		}
+		right_parts = append(right_parts, sub2.String())
+		rootCIDRNet = sub1
+	}
+	var resultPrefixes []string = nil
+	resultPrefixes = append(resultPrefixes, prefixes[:max_prefix_idx]...)
+	resultPrefixes = append(resultPrefixes, reverse(right_parts)...)
+	resultPrefixes = append(resultPrefixes, prefixes[max_prefix_idx+1:]...)
+
+	// return result
+	return rootCIDRNet.String(), resultPrefixes, nil
+}
+
+func removeDuplicates(elements []string) []string {
+	encountered := map[string]bool{}
+	result := []string{}
+
+	for v := range elements {
+		if !encountered[elements[v]] {
+			encountered[elements[v]] = true
+			result = append(result, elements[v])
+		}
+	}
+	return result
+}
+
+func reverse(values []string) []string {
+	newValues := make([]string, len(values))
+
+	for i, j := 0, len(values)-1; i <= j; i, j = i+1, j-1 {
+		newValues[i] = values[j]
+		if i != j {
+			newValues[j] = values[i]
+		}
```

Comment:
if i == j {
        continue
    }
    newValues[i], newValues[j] = values[j], values[i]

Reply:
Thank you, fixed

Reply:
Marked this comment as resolved, by mistake. Suggested loop changes semantics. 
`newValues[i] = values[j]` should be executed in any case, not only if `i != j`

Reply:
exactly, because we also copying values into another place :) Comment updated.

Reply:
Now it works as intended, thank you, fixed